### DISCUSSION TITLE
NodeJS Client: Cloudflare worker compatibility?

### DISCUSSION THREAD

**Discussion by trevhunter** [ravendb/ravendb#16465] [Q&A] [answered]
The cloudflare worker's platform runs a custom runtime of NodeJS with a [few restrictions](https://developers.cloudflare.com/workers/platform/nodejs-compatibility/) such as limited access to some API's like filesystem etc.

It seems the [RavenDB NodeJs](https://github.com/ravendb/ravendb-nodejs-client) client isn't quite compatible with these restrictions at the moment as it imports some of the restricted API's, resulting in some build errors when deploying a cloudflare worker or cloudflare pages function that uses the node.js client:

```
✘ [ERROR] Could not resolve "util"

    ../node_modules/verror/lib/verror.js:6:23:
      6 │ var mod_util = require('util');
```

It would be really nice to see RavenDB [officially supported as a database on Cloudflare](https://developers.cloudflare.com/workers/learning/integrations/databases/) but in the meantime, Is there a cut down version of the NodeJS client that could be compatible with running in a Cloudflare workers environment?

**Comment by trevhunter** [comment]
I did notice https://github.com/ravendb/ravendb-nodejs-client/pull/360, and was wondering if you have any working examples of the RavenDB client in Cloudflare workers?

**Comment by kamranayub** [comment]
@trevhunter Good news, we've got a [template created](https://github.com/ravendb/template-cloudflare-worker), [docs how-to PR](https://github.com/ravendb/docs/pull/1616) in progress, and [a walkthrough video](https://f.io/MiwP4KQb) in final draft.

The key bit is to add `node_compat = true` to your Wrangler config. This makes the SDK compatible.

It all works, it's just not fully live yet -- I just made the template live 😄 The docs are pretty much finalized, there's maybe only some minor QA stuff to do. The video link will let you access and watch the video, again it's pretty final except for minor QA work.

Feel free to leave feedback here, in the PR, or on the video with comments.

Once everything is final and live, I'll be sure to request RavenDB to be added to that integrations page 👍 

cc @ml054

**Reply by trevhunter** [reply]
This is absolutely fantastic @kamranayub ! Thanks for this goodness.

The template makes it super easy to get started, but I'd also suggest some instructions on how to manually add things to an existing Workers project. I'll be trying that (based on what I see the template doing) and will try and take notes.

**Reply by trevhunter** [reply]
So today I was able to give the template a go and things seem pretty good - up until the 2nd request. Here's what I did.

1. I followed the instructions, set the template up, got it working with my own instance on RavenDB cloud (things show connected and green on the welcome page). Nicely done on this!
2. I added a small handler to get a document by ID and return the json to the client. This works fine exactly once, after which I get the following exception:

```
RavenException: An exception occurred while contacting https://[someserver].ravendb.cloud/databases/[somedb]/docs?id=test: Cannot read properties of null (reading 'getReader') . 
```

Here's the code in the handler:
``` ts
// Gets a document from the db and returns the JSON
router.get('/docs/:id', docGetHandler);

async function docGetHandler(request: IRequest, env: Env) {

	let id = request.params.id;
	const session = env.db;
	const doc = await session.load(id); // works once, blows up on 2nd call.
	if (!doc) {
		return new Response('404, not found! No document with id: "' + id + '"', { status: 404 });
	}
	return new Response(JSON.stringify(doc), {
		status: 200,
		headers: { 'content-type': 'application/json' },
	});
}
```
The above handler works for the first request - I get the document json back in the browser. But any subsequent requests fail with the above error. It seems like something is getting corrupted in the document store after that first request (or perhaps I'm misusing the nodejs client in some way?)

**Reply by ml054** [reply]
@trevhunter 
Here it is:
https://github.com/ravendb/ravendb-nodejs-client/issues/365

It wasn't released yet. 

We will release next version with this fix.

**Reply by ml054** [reply]
node.js client 5.2.9 was relased. Please give a try with that version.

**Reply by trevhunter** [reply]
@ml054 I just tried with 5.2.9 - same result - it throws that exception on the 2nd request to `load()`.

A few more specific notes:
1. It only seems to happen when loading the **same** document again (i.e. the document was previously loaded in the document store). Loading document "docs/1" and then "docs/2" works fine, but if I attempt to load either of those documents again, it gives that error.
2. Looking in Traffic Watch on the RavenDB side: The first request gets a `HTTP 200` response. The 2nd request gets a `HTTP 304` response. 

![image](https://github.com/ravendb/ravendb/assets/3181772/b92e85fa-a126-47ca-8648-259adda9fa8a)
 

This seems to indicate to me something is up with client side caching of the entities and that the node client isn't handling the 304 correctly. So I modified the code a bit to disable client side caching and it seems to work fine:

``` ts
const dbSession = dbStore.openSession({noCaching: true});
```

I can get by with the no caching workaround for now, but hopefully this helps track down the issue for you.

**Reply by ml054** [reply]
ok Now I get it. 

Please use noCaching as temporary workaround. 

I've opened https://issues.hibernatingrhinos.com/issue/RDBC-715/304-response-code-isnt-handled-properly-in-cloudflare for investigation. 

I'm sorry for inconvience and thanks for sharing that with us.

**Reply by kamranayub** [reply]
Thanks @trevhunter, appreciate you giving it a go! I will add a TODO to add docs around adding support to an existing project. 

One additional thought @ml054 I didn't bring up yet but will now is it seems like there's a possibility to take advantage of [KV](https://developers.cloudflare.com/workers/runtime-apis/kv/) (or maybe Durable Objects, though that's paid-only), especially for aggressive caching. But it would require quite a bit more testing/research. Perhaps it would be a separate package or something since it would be specific to CF platform.

**Reply by trevhunter** [reply]
No problem at all. Thanks for the support. I'll keep testing and try to log any reproducible bugs I find. 

Both Node and Cloudflare are a bit of an experiment for me at the moment, but hopefully I'll be able to make it production worthy soon enough.

**Reply by ml054** [reply, marked as answer]
@trevhunter 

I've managed to reproduce your case. Looks like currently cloudflare allows to initialize document store outside of request handling function. :)

Issue with 304 issues was fixed in 5.2.10 release. Please update your dependency to latest version.

**Reply by trevhunter** [reply]
Thanks @ml054 - 5.2.10 looks good without `nocaching`. I'll keep an eye out for anything more under CF workers.

**Reply by ml054** [reply]
Glad to hear that. Good luck. 

Note: In general the most tricky part about node.js vs cloudflare is that node uses stream, and fetch function from cloudflare uses WebStreams (https://nodejs.org/api/webstreams.html). Interation is mostly about connecting those 2 little bit different worlds. :)

**Reply by trevhunter** [reply]
Thanks for the info.

In case it's helpful, I did notice that Cloudflare just launched [support for NodeJS streams last week as part of their developer week](http://blog.cloudflare.com/workers-node-js-apis-stream-path/) (among other APIs).

**Comment by kamranayub** [comment]
Hi all, we've released both the how-to guide and a template for Cloudflare:

- https://ravendb.net/docs/article-page/5.4/csharp/start/guides/cloudflare-workers/overview
- https://github.com/ravendb/template-cloudflare-worker
